### ISSUE TITLE
How to fix "OpenSSL failed - error:0A000086:SSL routines::certificate verify failed"

### ISSUE THREAD

**Issue by peterphonic** [machinezone/IXWebSocket#421]
Hello.  Sorry to post here, but maybe you will be able to tell we what to look to fix my problem.

### Details
I am using the IXWebSocket client example and I try to open a connection to a **nodejs https/websocket server** but the client refuses to connect and returns the error : _OpenSSL failed - error:0A000086:SSL routines::certificate verify failed_

### What I tried
I tried many things:

1. I have a React front end using wss websocket to communicate to my https nodejs server. **This works**.

2. I compiled IXWebSocket without openSSL.   The IXWebSocket client without tlsOptions, is connecting to my same nodejs server, http/websocket (non secure). **It works**.
3. I compiled IXWebSocket with openSSL.  It was a bit more challenging.  I followed the instructions to compile OpenSSL, and then I used scons to compile IXWebSocket, with IXWEBSOCKET_USE_OPEN_SSL and IXWEBSOCKET_USE_TLS flags enable.  Everything compile.  Then I compiled the little example client with ixwebsocket.lib, libssl.lib,libcrypto.lib,Ws2_32, Crypt32, and Shlwapi.  Everything looks fine, but the client returns the error mention above.  I followed the instruction [here](https://github.com/machinezone/IXWebSocket/issues/386#issuecomment-1105235227) to set the tlsOption. 

```
   ix::SocketTLSOptions tlsOptions;
    tlsOptions.certFile = "WebRTC.test.crt";
    tlsOptions.keyFile = "WebRTC.test.key";
    tlsOptions.caFile = "WebRTC-CA.pem";   //Also, I am not sure the .pem is necessary here. 
    webSocket.setTLSOptions(tlsOptions);
    std::string url("wss://localhost:8080");
    //std::string url("wss://echo.websocket.org");
    webSocket.setUrl(url);
```

Because number one above works, I think my key and my certificate are correct..  Would it be possible that the problem is the result of the compilation of openSSL is incorrect and IXWebSocket fails because of that?   

### EDIT
Here is my command line to build the websocket client
```
cl /Fobuild\fwk\webrtc\testWebSocketClient\MSVC14.1\x86_64\debug\testWebSocketClient.obj /c fwk\webrtc\testWebSocketClient\testWebSocketClient.cpp /TP /FC /nologo /Od /D_DEBUG /EHsc /RTC1 /MTd /IC:\SVN\3rdParty\3rdPartyPackages\IXWebSocket-11.4.3 /Z7 /Fdbuild\fwk\webrtc\testWebSocketClient\MSVC14.1\x86_64\debug\testWebSocketClient.obj.pdb
testWebSocketClient.cpp
link /INCREMENTAL /DEBUG /NOLOGO /MACHINE:X64 /OUT:build\fwk\webrtc\testWebSocketClient\MSVC14.1\x86_64\debug\testWSClient.exe /LIBPATH:C:\SVN\3rdParty\3rdPartyPackages\IXWebSocket-11.4.3\MSVC14.1\x86_64\debug /LIBPATH:C:\SVN\3rdParty\3rdPartyPackages\openssl-3.0.7 Ws2_32.lib Crypt32.lib Shlwapi.lib ixwebsocket.lib libssl.lib libcrypto.lib /PDB:build\fwk\webrtc\testWebSocketClient\MSVC14.1\x86_64\debug\testWSClient.exe.pdb /DEBUG build\fwk\webrtc\testWebSocketClient\MSVC14.1\x86_64\debug\testWebSocketClient.obj
LINK : build\fwk\webrtc\testWebSocketClient\MSVC14.1\x86_64\debug\testWSClient.exe not found or not built by the last incremental link; performing full link
```
### EDIT 2
I am wondering if it is normal that SSL_OP_NO_TLSv1_3 is defined for me.  Who is setting this define?  
```
#ifdef SSL_OP_NO_TLSv1_3
            // (partially?) work around hang in openssl 1.1.1b, by disabling TLS V1.3
            // https://github.com/openssl/openssl/issues/7967
            options |= SSL_OP_NO_TLSv1_3;
#endif
            SSL_CTX_set_options(ctx, options);
        }
        return ctx;
    }
```

Any ideas how to debug this is welcome.  Thank you.

**Comment by peterphonic** [comment]
I looked deeper in the OpenSSL code, and for a reason I don't know, a fatal error is triggered.  In the code below, I have no idea yet  why  _s->verify_mode_ is equal to _SSL_VERIFY_PEER_ and _i_ equal to _0_.  Any idea are welcome1


![image](https://user-images.githubusercontent.com/5274077/208190314-500f01b3-4f69-49ed-9ffd-80f46c9d0050.png)

**Comment by peterphonic** [comment]
Thanks for your reply,

I am really new in the TLS world, and it is not yet clear to me where this logic is.  This is a silly question, but TLS logic is implemented by `IXWebSocket` or by `OpenSSL` ?

> I had problem with openssl, inside TLS-1.3 so I disabled it explicitely (when a version of openssl is >= 3.0, or support tls-1.3, that macro is defined).

I confirmed that I built `IXWebSocket` with OpenSSL 3.0.7

- In other word, that means that `IXWebSocket` cannot use TLS 1.3 at all?  Could this be a problem?
- Is this something that can be fixed in `IXWebSocket` or the fix has to be in `OpenSSL`?
- Should I roll back to `OpenSSL` 1.1.1 instead? 

> You will need to use the same keys on both client and server (your node.js one).

Something I realized, my key/certificates were generated with `OpenSSL` 1.1.1, and I use `IXWebSocket` with `OpenSSL` 3.07, that might be a problem. 

Regards

**Comment by bsergean** [comment]
There are no silly questions :)

TLS logic is implemented by IXWebsocket, which calls into OpenSSL.

If you want to encrypt the connection, but skip validating certificates, you could do this:

```
tlsOptions.caFile = "NONE";
```

I believe that your error is because you are not passing the same certificates in your client and your servers.

**Comment by peterphonic** [comment]
Regarding TLS 1.3 not supported in `IXWebSocket`

1. Does this have any implications?  Like if my client wants to talk with a server that has TLS 1.3 implemented? My websocket server is in `nodejs` 18, which has his own TLS implementation, I don't think it relies on OpenSSL.
2. In order to have TLS 1.3 working in `IXWebSocket`, a fix would be needed in OpenSSL or in `IXWebSocket`?

Regards

**Comment by bsergean** [comment]
We could try to remove the #ifdef in ixwebsocket, maybe the TLS 1.3 problem is fixed now.

On your node.js server you could say “only allow tls-1.2 probably”. And node.js uses openssl under the hood I think.

**Comment by peterphonic** [comment]
> We could try to remove the #ifdef in ixwebsocket, maybe the TLS 1.3 problem is fixed now

I'll try it when I will be back from vacation in January.  I`ll let you know for sure.  We are in the start of a new project here, we are working on creating https servers and websocket servers from scratch and our client will for sure asks for the last security updates, so chances are good that TLS 1.3 will be mandatory.

> And node.js uses openssl under the hood I think.

You are totally right, just did a quick search and nodejs website says : _Node.js v17.x, v18.x, and v19.x use OpenSSL v3_.

Regards

**Comment by peterphonic** [comment]
Hello,

So, everything is fine.  I originally created my key/certificate with OpenSSL 1.1.1.  I re-created my key/certificate with OpenSSL 3.07 and everything is working fine!  My IXWebSocket client can communicate with my secured javascript websocket server.

Also, I removed the `#ifdef SSL_OP_NO_TLSv1_3` from IXSocketOpenSSL.cpp, and my client is sill working well.  What kind of test is needed to make sure I did not break something?
